modules/standard/Regex.chpl

```diff
@@ -611,12 +619,65 @@ record regex {
 
   proc init=(x: regex(?)) {
     this.exprType = x.exprType;
-    this.home = x.home;
-    this._regex = x._regex;
-    this.complete();
-    on home {
-      qio_regex_retain(_regex);
+    /* always bring the regex local */
+    this.home = here;
+    /* if it's local, retain and avoid the recompile (thread safe) */
+    if (x.home == here) {
+      this._regex = x._regex;
+      this.complete();
+      qio_regex_retain(x._regex);
+    } else {
+      /* otherwise recompile locally */
+      this.complete();
+      var serialized = x._serialize();
+      this._deserialize(serialized);
+    }
+  }
+
+  pragma "no doc"
+  proc _serialize() {
+    var pattern: exprType;
+    var options: qio_regex_options_t;
+    var _regexCopy = _regex;
+
+    on this.home {
+      /* NOTE we can't reference `this` inside this on block because
+       * it can cause a recursive chpl__serialize loop */
+      var patternTemp: c_string;
+      qio_regex_get_pattern(_regexCopy, patternTemp);
+      if exprType == string then {
+        try! pattern = createStringWithOwnedBuffer(patternTemp);
+      }
+      else {
+        pattern = createBytesWithOwnedBuffer(patternTemp);
+      }
+
+      var localOptions: qio_regex_options_t;
+      qio_regex_get_options(_regexCopy, localOptions);
+      options = localOptions;
     }
+    return new chpl_serializeHelper(exprType, pattern, options);
+  }
+
+  pragma "no doc"
+  proc _deserialize(in data) {
```

Comment:
The `in` intent here doesn't seem right. I think it would cause a copy of the object to be made inside of `chpl__deserialize` before calling this routine, because `chpl__deserialize` passes the argument by `const ref`.

Reply:
Oh, I see, that would only lead to copying the string data in the `chpl_serializeHelper`. Even still, it's probably better to not put `in` here. Is there a particular reason it is here?

Reply:
So we took a look at this and found that the `extern qio_regex_create_compile` was missing a `const` on `const ref options:qio_regex_options_t`. This fixes the error but lead to having to call `text.localize()` in `_deserialize`, which shouldn't happen because all callers should already have a local string. This appears to be because our `chpl__serialize` doesn't cause recursive `chpl__serialize` on the string/bytes! This could fit well into the next regex effort in reducing allocations and/or comms so I'm wondering if you and/or @ronawho think we could merge this as is and fix it in our next effort (which is in my current sprint todo).

Reply:
Merging it as-is is fine with me

Reply:
(bit at least fix the const ref in that extern qio_regex_create_compile)

Reply:
fixed the const ref. `localize` is necessary for correctness and should be removed in near future work